src/ast/passes/semantic_analyser.cpp

```diff
@@ -4077,7 +4076,32 @@ void SemanticAnalyser::validate_new_key(const SizedType &current_key_type,
         valid = false;
       }
     } else if (!current_key_type.IsEqual(new_key_type)) {
-      valid = false;
+      if (current_key_type.IsIntegerTy()) {
+        auto *integer = key_expr.as<Integer>();
+        if (integer) {
+          uint64_t value = integer->value;
+          bool can_fit = false;
+          if (!current_key_type.IsSigned()) {
+            auto min_max = getUIntTypeRange(current_key_type);
+            can_fit = value <= min_max.second;
+          } else {
+            auto min_max = getIntTypeRange(current_key_type);
+            can_fit = value <= static_cast<uint64_t>(min_max.second);
```

Comment:
I'm confused. For the signed case, why are we not also looking at the minimum value?

Reply:
For example, this isn't expected, right?

```
$ sudo ./build/src/bpftrace -e 'BEGIN { let $x: int8 = -5; $x = -99999; print($x) }'
[sudo] password for dxu:
Attaching 1 probe...
-99999
^C
```

Reply:
Map keys and values are automatically up-casted to 64 bit ints so even if you cast a value to a smaller int, it still sets it to 64 bits. Also your example is for variables.

Reply:
Ah right, there's the NegativeInteger AST node now. Never mind. 

But I was showing the scratch var example b/c the code paths look similar. It's a bug above, right? 

Also relevant since I know about NegativeInteger node, should that be considered in these checks (separate PR territory)

Reply:
> It's a bug above, right?

Yeah, looks like it. I'll look into that.

> Also relevant since I know about NegativeInteger node, should that be considered in these checks

We actually don't have to worry about that at this point because of the underflow checks that happen earlier (and because of the 64 bit upcasting).